src/board/v3/init.c

```diff
@@ -55,38 +63,68 @@ static void reset_callback_register(void)
   );
 }
 
-void init(void)
+void pre_init(void)
 {
   led_setup();
+}
 
+void init(void)
+{
+  fault_handling_setup();
+  reset_callback_register();
+
+  nap_version_check();
   nap_setup();
+  nap_auth_check();
+  nap_callbacks_setup();
 
   srand(0);
+}
 
-  usarts_setup();
-  s32 serial_number = -1;//nap_conf_rd_serial_number();
-  if (serial_number < 0) {
-    /* TODO: Handle this properly! */
-    serial_number = 0x2222;
-  }
-  sbp_setup(serial_number);
-
-  fault_handling_setup();
-
-  nap_callbacks_setup();
+static bool nap_version_ok(u32 version)
+{
+  return ((version & REQUIRED_NAP_VERSION_MASK) == REQUIRED_NAP_VERSION_VAL);
+}
 
-  reset_callback_register();
+static void nap_version_check(void)
+{
+  u32 nap_version = nap_conf_rd_version();
+  if (!nap_version_ok(nap_version)) {
+    while (1) {
+      log_error("Unsupported NAP version: 0x%08x", nap_version);
+      chThdSleepSeconds(2);
+    }
+  }
 }
 
 /* Check NAP authentication status. Block and print error message
  * if authentication has failed. This must be done after the NAP,
  * USARTs, and SBP subsystems are set up, so that SBP messages and
  * be sent and received (it can't go in init() or nap_setup()).
  */
-void check_nap_auth(void)
+static void nap_auth_check(void)
 {
+
```

Comment:
In the latest NAP version, `NAP->STATUS` bit 0 is 1 as long as the NAP is locked.